components/brave_wallet/browser/eth_abi_decoder.cc

```diff
@@ -0,0 +1,306 @@
+      value = GetAddressArrayFromData(data, offset);
+    } else {
+      // For unknown/unsupported types, we only extract 32-bytes. In case of
+      // dynamic types, this value is a calldata reference.
+      value = GetArgFromData(data, offset);
+    }
+
+    if (!value)
+      return absl::nullopt;
+
+    // On encountering a dynamic type, we extract the reference to the start
+    // of the tail section of the calldata.
+    if ((type == "bytes" || type == "string" || base::EndsWith(type, "[]")) &&
+        calldata_tail == 0) {
+      auto pointer = GetSizeFromData(data, offset);
+      if (!pointer)
+        return absl::nullopt;
+
+      calldata_tail = *pointer;
+    }
+
+    offset += 32;
+
+    tx_args.push_back(*value);
+    tx_params.push_back(type);
+  }
+
+  // Extraneous calldata, unless in the tail section, is considered invalid.
+  if (offset != calldata_tail && offset < data.size())
+    return absl::nullopt;
+
+  return std::make_tuple(tx_params, tx_args);
+}
+
+}  // namespace
+
+// UniswapEncodedPathDecode parses a Uniswap-encoded path and return a vector
+// of addresses representing each hop involved in the swap.
+//
+// Single-hop swap: Token1 → Token2
+// Multi-hop swap: Token1 → Token2 → WETH → Token3
+//
+// Each encoded hop contains a 3-byte pool fee, which is associated with the
+// address that follows. It is excluded from the result of this function.
+//
+// ┌──────────┬──────────┬──────────┬─────┐
+// │ address  │ pool fee │ address  │     │
+// │          │          │          │ ... │
+// │ 20 bytes │ 3 bytes  │ 20 bytes │     │
+// └──────────┴──────────┴──────────┴─────┘
+absl::optional<std::vector<std::string>> UniswapEncodedPathDecode(
+    const std::string& encoded_path) {
+  std::vector<uint8_t> data;
+  if (!PrefixedHexStringToBytes(encoded_path, &data))
+    return absl::nullopt;
+  size_t offset = 0;
+  std::vector<std::string> path;
+
+  // The path should be long enough to encode a single-hop swap.
+  // 43 = 20(address) + 3(fee) + 20(address)
+  if (data.size() < 43)
+    return absl::nullopt;
+
+  // Parse first hop address.
+  path.push_back("0x" + HexEncodeLower(data.data(), 20));
+  offset += 20;
+
+  while (true) {
+    if (offset == data.size())
+      break;
+
+    // Parse the pool fee, and ignore.
+    if (data.size() - offset < 3)
+      return absl::nullopt;
+
+    offset += 3;
+
+    // Parse next hop.
+    if (data.size() - offset < 20)
+      return absl::nullopt;
+    path.push_back("0x" + HexEncodeLower(data.data() + offset, 20));
+    offset += 20;
+  }
+
+  // Require a minimum of 2 addresses for a single-hop swap.
+  if (path.size() < 2)
+    return absl::nullopt;
+
+  return path;
+}
+
+absl::optional<std::tuple<std::vector<std::string>,   // tx_params
+                          std::vector<std::string>>>  // tx_args
+ABIDecode(const std::vector<std::string>& types,
+          const std::vector<uint8_t>& data) {
+  return ABIDecodeInternal(types, data);
```

Comment:
can we eliminate the extra layer of calling `ABIDecodeInternal`?

Reply:
Done 8cfe65bfe4338bf0bd4ce47e4b0cfff6a1ef69d4.